packages/steps/src/BuildStepOutput.ts

```diff
@@ -1,39 +1,55 @@
+import assert from 'assert';
+
 import { BuildStepContext } from './BuildStepContext.js';
 import { BuildStepRuntimeError } from './errors/BuildStepRuntimeError.js';
 
 export class BuildStepOutput {
   public readonly id: string;
-  public readonly stepId: string;
+  public readonly stepId?: string;
```

Comment:
I'm not to familiar with that code, so I might be missing context, but this seems to be like a hack. 

If I understand correctly you are allowing undefined here because the function itself does not have a step id until it's called somewhere. and the latter clone is creating a proper build step with a step id.

It seems to me that a better approach would be if BuildFunction was holding BuildStepInputProvider and BuildStepOutput providers that would be used to create BuildStepOutput and BuildStepInput when creating BuildStep. Also, the clone method name is quite confusing there.